omc3/kmod_importer.py

```diff
@@ -0,0 +1,291 @@
+    meas_paths: Sequence[Path | str],
+    df_model: tfs.TfsDataFrame,
+    beam: int,
+    output_dir: Path | str,
+    show_plots: bool = False,
+    ) -> dict[str, dict[int, tfs.TfsDataFrame]]:
+    """ Averages all kmod results.
+
+    Args:
+        meas_paths (Sequence[Path | str]): Paths to the K-modulation results. 
+        df_model (tfs.TfsDataFrame): DataFrame with the model. 
+        beam (int): Beam for which to average. 
+        output_dir (Path | str, optional): Path to the output directory. Defaults to None.
+        show_plots (bool, optional): If True, show the plots. Defaults to False.
+
+    Returns:
+        dict[int, tfs.TfsDataFrame]: Averaged kmod results, sorted by IP. 
+    """
+    sorted_paths = _sort_paths_by_ip(meas_paths, beam)
+
+    averaged_results = {}
+    for ip, paths in sorted_paths.items():
+        LOG.debug(f"Averaging IP {ip}")
+        betastar = [round(bstar, 2) for bstar in df_model.loc[ip, [f"{BETA}X", f"{BETA}Y"]]]
+
+        average = average_kmod_results(
+            ip=int(ip[-1]),
+            betastar=betastar,
+            meas_paths=paths,
+            output_dir=output_dir,
+            plot=True,
+            show_plots=show_plots
+        )
+        averaged_results[ip] = average
+
+    return averaged_results
+
+
+def _sort_paths_by_ip(paths: Sequence[str | Path], beam: int) -> dict[str, list[str | Path]]:
+    """ Sorts the kmod results files by IP. 
+    
+    Identification of the IP is done by reading the `lsa_results.tfs` files.
+    """
+    sorted_paths = defaultdict(list)
+    for path in paths:
+        path = Path(path)
+        beam_dir = path / f"{BEAM_DIR}{beam}"
+        lsa_results = beam_dir / f"{LSA_FILE_NAME}{EXT}"
+        df = tfs.read(lsa_results, index=NAME)
+        ip = [ip.upper() for ip in IPS if ip.upper() in df.index][0]
+        sorted_paths[ip].append(path)
+    return sorted_paths
+
+
+# Lumi Imbalance ---
+
+def calculate_all_lumi_imbalances(
+    averaged_results: dict[str, dict[int, tfs.TfsDataFrame]], 
+    output_dir: Path | str = None
+    ) -> None:
+    """ Calculates the luminosity imbalance between two IPs.
+    
+    Args:
+        averaged_results (dict[str, dict[int, tfs.TfsDataFrame]]): Averaged kmod results, sorted by IP. 
+        output_dir (Path | str, optional): Path to the output directory. Defaults to None.
+
+    Returns:
+        tfs.TfsDataFrame: DataFrame with the luminosity imbalance.
+    """
+    sets_of_ips = list(combinations(averaged_results.keys(), 2))
+    for (ipA, ipB) in sets_of_ips:
+        LOG.debug(f"Calculating lumi imbalance between {ipA} and {ipB}")
+
+        # Calculate luminosity imbalance
+        data = {ip.lower(): averaged_results[ip][0] for ip in (ipA, ipB)}
+        try:
+            df = calculate_lumi_imbalance(**data, output_dir=output_dir)
+        except KeyError as e:
+            LOG.debug(f"Could not calculate lumi imbalance between {ipA} and {ipB}. Skipping.", exc_info=e)
```

Comment:
This was on purpose, because you might import data that is not averaged data, which in turn does not have a BEAM column or you import averaged data, but only for one beam. As the lumi imbalance calculation is only a "side effect" of the importing, I think it is OK to go with debug here.

Reply:
Right, that's fair enough. Maybe a comment above the code stating that?